modules/gapi/src/compiler/passes/match_pattern_to_substitute.cpp

```diff
@@ -0,0 +1,106 @@
+using Graph = GModel::Graph;
+using Metadata = typename Graph::CMetadataT;
+using VisitedMatchings = std::list<std::pair<ade::NodeHandle, ade::NodeHandle>>;
+
+// Returns true if two DATA nodes are semantically identical:
+//  - both nodes have the same GShape
+//  - both nodes have the same storage
+//
+// @param firstMeta - metadata of first
+// @param secondMeta - metadata of second
+bool compareDataNodes(const Metadata& firstMeta,
+                      const Metadata& secondMeta)
+{
+    GAPI_Assert(firstMeta.get<NodeType>().t == NodeType::DATA);
+    GAPI_Assert(firstMeta.get<NodeType>().t == secondMeta.get<NodeType>().t);
+
+    const auto& firstData = firstMeta.get<Data>();
+    const auto& secondData = secondMeta.get<Data>();
+    // compare shape
+    if (firstData.shape != secondData.shape) {
+        return false;
+    }
+    // compare storage
+    if (firstData.storage != secondData.storage) {
+        return false;
+    }
+
+    // FIXME: is it enough to only check shape && storage?
+    return true;
+}
+
+// Returns matched pairs of {pattern node, substitute node}
+SubgraphMatch::M matchDataNodes(const Graph& pattern,
+                                const Graph& substitute,
+                                const std::vector<ade::NodeHandle>& patternNodes,
+                                std::vector<ade::NodeHandle> substituteNodes)
+{
+    SubgraphMatch::M matched;
+    for (auto it : ade::util::zip(patternNodes, substituteNodes)) {
+        const auto& pNode = std::get<0>(it);
+        const auto& sNode = std::get<1>(it);
+        if (!compareDataNodes(pattern.metadata(pNode), substitute.metadata(sNode))) {
+            return {};
+        }
+        matched.insert({ pNode, sNode });
+    }
+    return matched;
+}
+}  // anonymous namespace
+
+SubgraphMatch matchPatternToSubstitute(const cv::gimpl::GModel::Graph& pattern,
+                                       const cv::gimpl::GModel::Graph& substitute,
+                                       const cv::gimpl::Protocol& patternP,
+                                       const cv::gimpl::Protocol& substituteP)
+{
+    //---------------------------------------------------------------
+    // Match data nodes which start and end our pattern and substitute
+    const auto& patternDataInputs = patternP.in_nhs;
+    const auto& patternDataOutputs = patternP.out_nhs;
+
+    const auto& substituteDataInputs = substituteP.in_nhs;
+    const auto& substituteDataOutputs = substituteP.out_nhs;
+
+    // if number of data nodes doesn't match, abort
+    if (patternDataInputs.size() != substituteDataInputs.size()
+        || patternDataOutputs.size() != substituteDataOutputs.size()) {
+        return {};
+    }
+
+    // for each pattern input we must find a corresponding substitute input
+    auto matchedDataInputs = matchDataNodes(pattern, substitute, patternDataInputs,
+        substituteDataInputs);
+    // if nothing found, abort
+    if (matchedDataInputs.empty()) {
+        return {};
```

Comment:
assert

Reply:
I wonder if it's better to return and empty match at this level and decide what to do with this in the caller's scope: e.g. assert or throw an exception with a proper explanation. I think, in theory, we can benefit from having this function returning `{}` instead of always asserting. `return {}` gives us more freedom for how to use the result